domains/pente/src/main/java/io/kaleido/paladin/pente/domain/PenteDomain.java

```diff
@@ -282,6 +285,17 @@ protected CompletableFuture<AssembleTransactionResponse> assembleTransaction(Ass
                      setAssemblyResult(AssembleTransactionResponse.Result.REVERT).
                      setRevertReason(e.getMessage()).
                      build());
+         } catch (ExecutionException e) {
+             if (e.getCause() instanceof ErrorResponseException) {
+                // Any error response from a plugin during assembly is considered a revert.
+                // These can stem from things like an invalid ABI or inputs.
+                LOGGER.error(new FormattedMessage("Error response from plugin during assemble for TX {}", request.getTransaction().getTransactionId()), e);
+                return CompletableFuture.completedFuture(AssembleTransactionResponse.newBuilder().
```

Comment:
oh - this is just using the spelling that already exists, but in the Java plugin.

Reply:
Not quite what I'm saying.

I think the only code that can determine if a failure is permanent, is the plugin.

So Paladin should have a spelling to tell the difference explicitly - **not** with PD codes, because those are out of its scope.

- Plugin returns any error that isn't explicitly a "please treat this as a revert": Keep retrying
- Plugin returns an explicit revert: This is a revert

I now understand you've put code into the the Java SDK - to try and guess the difference, and translate that into the Paladin SDK. I disagree with that approach.

I think instead the Java SDK should force the Java plugin to do the right thing.

That could be by defining a specific type of Class of Exception that the code that throw to explicitly revert.

Reply:
Thanks Andrew, this is super helpful. Yes 👍

Reply:
I've updated this interface with an "error_type" that currently is UNKNOWN or INVALID_INPUT. The encode/decode operations have been adjusted to return INVALID_INPUT in (I think) appropriate cases, and Pente is adjusted to interpret _only this_ error code as a revert.

Reply:
Note that there is an earlier `catch` specifically for EVM execution errors, which was already using the "revert" behavior. This is just adding another `catch` for this other category of failure.